Add simple external test covering cache verification

josvazg opened the pull request:
Signed-off-by: Jose Luis Vazquez Gonzalez <josvaz@vmware.com>

This is the simplest way I can think of we ensure that, at the very least, if the cache stops working as expected we will notice in our testing. Deeper testing would be a lot more costly in time to what I can invest right now.

josvazg commented:
I should add a small sub test to ensure a second usage of the layer uses the cached file this time.

petewall commented on test/external/external_test.go:
```diff
@@ -224,13 +255,25 @@ var _ = Describe("External tests", func() {
 
-		define.Then(`^remove the archive folder at testchart-intermediate.tar$`, func() {
-			err := os.Remove("testchart-intermediate.tar")
+		define.When(`^removing relok8s-save-cache$`, func() {
+			err := os.RemoveAll(saveCacheFolder())
+			Expect(err).ToNot(HaveOccurred())
+		})
```
If this is just a cleanup step, you can add a second `func()` that runs after all of the regular steps.

http://www.bunniesandbeatings.com/goerkin/#cleanup-steps

josvazg replied:
will do

josvazg replied:
Actually this is not really a clean up. I will remove this to `clear relok8s-save-cache` because this is a pre-step before testing the cache. This ensures the first run is not using cached content, yet the cached content gets populated:
~~~go
		steps.When("clear relok8s-save-cache")
		steps.When(fmt.Sprintf("running relok8s chart move -y ../fixtures/testchart --image-patterns ../fixtures/testchart.images.yaml --to-intermediate-bundle %s/testchart-intermediate.tar", tmpDir))
...
		steps.Then("relok8s-save-cache contains expected layer")
~~~

Still, I did not know the cleanup trick for `goerkin`, and it might be useful elsewhere.

josvazg commented:
Abandoned in favor of #119

josvazg commented:
#119 got rejected so we resume this again to have test coverage of the caching support.

migmartri reviewed (changes requested):
Please don't update the headers of files that have not been updated.

did you get a error when running the linter locally or in the CI? note that I changed the github action to only take into account what has changed, locally it's by default like that.

josvazg commented:
> Please don't update the headers of files that have not been updated.
> 
> did you get a error when running the linter locally or in the CI? note that I changed the github action to only take into account what has changed, locally it's by default like that.

Done, I have removed the header renames also from this PR. I rename the only touched file here instead.

migmartri commented on test/external/external_test.go:
```diff
@@ -19,6 +19,15 @@ import (
 )
 
+func saveCacheFolder() string {
+	return filepath.Join(os.TempDir(), "relok8s-save-cache")
```
This is actually deleting the cache that's being used by relok8s itself right? I don't like that running a test has the side-effect of deleting the actual tooling cache.

Could we instead run these tests pointing to an unique, per-test cache path? This might require adding support for our tool via ENV variable or flag to point to the cache directory.

josvazg replied:
Do not want to add more code to be able to test, we already abandoned a change for that.

Maybe I can remove the specific layer instead.

migmartri replied:
nah, don't worry about it

josvazg replied:
Done

migmartri commented on test/external/external_test.go:
```diff
@@ -75,13 +84,35 @@ var _ = Describe("External tests", func() {
+		modtime := info.ModTime()
+		steps.When(fmt.Sprintf("running relok8s chart move -y ../fixtures/testchart --image-patterns ../fixtures/testchart.images.yaml --to-intermediate-bundle %s/testchart-intermediate-2.tar", tmpDir))
+		steps.And("the move is computed")
+		steps.Then("the command says it will archive the chart")
```
Is there much value on again checking the output of the command? Could we just check on the second run that it does not expose any error and that the layer is there?

josvazg replied:
Can remove

josvazg replied:
Done

migmartri commented on test/external/external_test.go:
```diff
@@ -75,13 +84,35 @@ var _ = Describe("External tests", func() {
+		info, err = os.Stat(expectedLayerFile())
+		Expect(err).ToNot(HaveOccurred())
+		Expect(info.ModTime()).To(Equal(modtime))
+		Expect(secondSaveDuration).To(BeNumerically("<", firstSaveDuration))
```
This test case can be brittle, what if extracting the file from local takes more than the network speed? It should not be the case in general but you are adding a dependency here to the underlying system.

josvazg replied:
timing checks removed

migmartri replied:
> --- FAIL: TestGroupChangesByChart (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
	
yep, i have it on my plate to take a look at it. I looked into it the other day but didn't manage to reproduce it

josvazg replied:
It is very weird indeed, it happened a few times to me for a little while, then did not happen again for some time. It looks like some kind of race condition somewhere.

migmartri reviewed (approved):
Thanks!

migmartri commented on test/external/external_test.go:
```diff
@@ -261,13 +283,25 @@ var _ = Describe("External tests", func() {
-		define.Then(`^remove the archive folder at testchart-intermediate.tar$`, func() {
-			err := os.Remove("testchart-intermediate.tar")
+		define.When(`^clear relok8s-save-cache expected layer$`, func() {
+			err := os.RemoveAll(expectedLayerFile())
```
nice, thanks!

migmartri reviewed (approved):
LGTM!
